src/core/core.tooltip.js

```diff
@@ -532,14 +542,15 @@ module.exports = function(Chart) {
 				tooltipSize = getTooltipSize(this, model);
 				alignment = determineAlignment(this, tooltipSize);
 				// Final Size and Position
-				backgroundPoint = getBackgroundPoint(model, tooltipSize, alignment);
+				backgroundPoint = getBackgroundPoint(model, tooltipSize, alignment, me._chart);
 			} else {
 				model.opacity = 0;
 			}
 
 			model.xAlign = alignment.xAlign;
 			model.yAlign = alignment.yAlign;
 			model.x = backgroundPoint.x;
+			model.xCaret = backgroundPoint.xCaret;
```

Comment:
After reading this comment I started thinking that why not to use caretX, maybe can avoid this xCaret complexity in code.  
And seems that [this current code](https://github.com/chartjs/Chart.js/blob/master/src/core/core.tooltip.js#L608) can be replaced with `x2 = vm.caretX;`.  
So I made new commit where I removed xCaret and made change described above.

In y axis direction tooltips seem to be working ok and no need for changes.

It's quite a lot commits here already. Should I rebase?